[Macros] Don't visit macro-generated extensions in `visitAuxiliaryDecls`.

hborla opened the pull request:
Macro-generated extensions are hoisted to file scope, because extensions are not valid in nested scopes. Callers of `visitAuxiliaryDecls` assume that the auxiliary decls are in the same decl context as the original, which is clearly not the case for extensions, and it leads to issues like visiting extension at the wrong time during SILGen. The extensions are already added exposed via `getTopLevelDecls()`, so we don't need to visit them as auxiliary decls.

Resolves: https://github.com/apple/swift/issues/66916, rdar://111000256

hborla committed:
[Macros] Don't visit macro-generated extensions in 'visitAuxiliaryDecls'.

Macro-generated extensions are hoisted to file scope, because extensions are
not valid in nested scopes. Callers of 'visitAuxiliaryDecls' assume that the
auxiliary decls are in the same decl context as the original, which is clearly
not the case for extensions, and it leads to issues like visiting extension at
the wrong time during SILGen. The extensions are already added to the top-level
decls, so we don't need to visit them as auxiliary decls, and we can type-check
macro-expanded decls at the end of visitation in TypeCheckDeclPrimary.

hborla commented:
@swift-ci please smoke test

DougGregor commented on lib/Sema/TypeCheckMacros.cpp:
```diff
@@ -1446,6 +1446,16 @@ swift::expandConformances(CustomAttr *attr, MacroDecl *macro,
+    // hoisting behavior of macro-generated extensions, we simply expose the
+    // extension to 'getTopLevelDecls()' and type check the extension eagerly
+    // here.
+    TypeChecker::typeCheckDecl(extension);
```
Could we instead visit these at the end of TypeCheckDeclPrimary's visitation? Doing a complete check of the extension seems unnecessary here.

(Also, you have typos "visitAuxiliarDecls" and "understsand").

hborla replied:
I suppose it's better to visit expanded extensions in TypeCheckDeclPrimary as opposed to visiting the top-level decls of `SF->getSynthesizedFile()` at the end of `TypeCheckSourceFileRequest::evaluate` because the former approach will definitely invoke the request if it hasn't already happened?

DougGregor replied:
I hadn't thought of `TypeCheckSourceFileRequest::evaluate`. Visiting the top-level decls of the synthesized file there should work.

hborla committed:
[Macros] Type-check macro-generated extensions via the top-level decls of
the synthesized source file instead of a separate API to gather the
extensions.

This is how macro-generated extensions are visited everywhere else.
